Add -version-constraints flag to upload subcommand

oliviermichaelis opened the pull request:
This PR adds a `-version-constraints` subcommand to the upload subcommand. Its intention is to fix #6 .
Let me know if there is a better place in the code for this functionality - I was thinking a lot about it but am not entirely happy about it...
Also let me know if I should a dedicated section in the `README.md`.

Thanks!

heimweh commented on internal/cli/uploadcmd/upload.go:
```diff
@@ -89,6 +104,7 @@ func New(config *rootcmd.Config) *ffcli.Command {
 	fs.StringVar(&cfg.S3Region, "s3-region", "", "Region of the S3 bucket when using the S3 registry type")
 	fs.StringVar(&cfg.GCSBucket, "gcs-bucket", "", "Bucket to use when using the GCS registry type")
 	fs.StringVar(&cfg.GCSPrefix, "gcs-prefix", "", "Prefix to use when using the GCS registry type")
+	fs.StringVar(&flagVersionConstraints, "version-constraints", "", "Limit the module versions that are eligible for upload with version constraints. The version string has to be formatted as a string literal containing one or more conditions, which are separated by commas")
```
Can we put this in the struct as well to be consistent with the other flags? 🙂 

```go
fs.StringVar(&cfg.VersionConstraints, ...)
```

oliviermichaelis replied:
That's what I did initially, but it didn't feel right to put a "meaningless" `string` in a struct. As essentially what we care about is the parsed and validated `version.Constraints`. So from an architecture perspective, that's what I want to pass on.
But I get your point, I spent a lot of time thinking which way I prefer. And how I did it doesn't feel correct as well, so I'm happy to change that :)

heimweh reviewed (commented):
Nice work @oliviermichaelis :tada:

This will be a great addition to the upload subcommand.

I left one minor comment.
Can you also add some example usage in the README on how to use it? 🙂  :pray:

oliviermichaelis commented:
So I renamed the flag to `-version-constraints-semver` and added another flag `-version-constraints-regex`. The second flag simply compiles a user-given regex and tries to match the version. If there is no match, the module is skipped.
Does that sound like a good idea? :)

heimweh commented on internal/cli/uploadcmd/upload.go:
```diff
@@ -70,10 +74,24 @@ func (c *Config) Exec(ctx context.Context, args []string) error {
 
+	// Validate the semver version constraints
+	if c.VersionConstraintsSemver != "" {
+		if _, err := version.NewConstraint(c.VersionConstraintsSemver); err != nil {
```
Can we store this constraint somewhere? Just so that we can use it directly inside `meetsSemverConstraints()` to avoid having to call NewConstraint inside that function? 🙂

oliviermichaelis replied:
Where would you want to store it, though? In the `Config` struct?
This is related to the comment that I raised here: https://github.com/TierMobility/boring-registry/pull/7#discussion_r644612132. So should we store both constraints in the `Config` alongside the flags? To me, it feels like we shouldn't pass on the flag variables once we've processed and parsed them. But it's also not that important, we just have to remember that the constraint is to be used, not the flag string variable.

heimweh replied:
Ah sorry, I meant the returned [constraints](https://pkg.go.dev/github.com/hashicorp/go-version#NewConstraint), not the string flag value :)

I see your point re: storing the strings that are only used once for parsing, 
let's declare them outside of the config, agree that it's a bit weird 👍

oliviermichaelis replied:
Sounds good, let's do that in a separate refactoring outside of the PR :)

oliviermichaelis replied:
Nevermind, just pushed another commit!:)

heimweh commented on internal/cli/uploadcmd/upload.go:
```diff
@@ -70,10 +74,24 @@ func (c *Config) Exec(ctx context.Context, args []string) error {
+
+	// Validate the regex version constraints
+	if c.VersionConstraintsRegex != "" {
+		if _, err := regexp.Compile(c.VersionConstraintsRegex); err != nil {
```
Same as above but for `meetsRegexConstraints()` 🙂

heimweh reviewed (commented):
Nice work @oliviermichaelis! Left two inline comments 🙂
